src/topology/inseparable.lean

```diff
@@ -178,12 +179,217 @@ def specialization_preorder : preorder X :=
+  stable_under_specialization (f ⁻¹' s) :=
+is_lower_set.preimage hs hf.specialization_monotone
+
+lemma stable_under_generalization.preimage {s : set Y}
+  (hs : stable_under_generalization s) (hf : continuous f) :
+  stable_under_generalization (f ⁻¹' s) :=
+is_upper_set.preimage hs hf.specialization_monotone
+
+/-- A map `f` between topological spaces is specializing if specializations lifts along `f`,
+i.e. for each `y ⬿ f x'` there is some `x ⬿ x'` whose image is `y`. -/
+def specializing (f : X → Y) : Prop :=
+relation.fibration (flip (⤳)) (flip (⤳)) f
+
+/-- A map `f` between topological spaces is generalizing if generalizations lifts along `f`,
+i.e. for each `y ⤳ f x'` there is some `x ⤳ x'` whose image is `y`. -/
+def _root_.generalizing (f : X → Y) : Prop :=
+relation.fibration (⤳) (⤳) f
+
+lemma specializing_iff_closure_singleton_subset :
+  specializing f ↔ ∀ x, closure {f x} ⊆ f '' closure {x} :=
+by simpa only [specializing, relation.fibration, flip, specializes_iff_mem_closure]
+
+alias specializing_iff_closure_singleton_subset ↔ specializing.closure_singleton_subset _
+
+lemma specializing.stable_under_specialization_image (hf : specializing f)
+  {s : set X} (hs : stable_under_specialization s) : stable_under_specialization (f '' s) :=
+is_lower_set.image_fibration hf hs
+
+alias specializing.stable_under_specialization_image ← stable_under_specialization.image
+
+lemma specializing_iff_image_singleton_stable_under_specialization :
+  specializing f ↔ ∀ x, stable_under_specialization (f '' closure {x}) :=
+by simpa only [closure_singleton_eq_Iic] using relation.fibration_iff_is_lower_set_image_Iic
+
+lemma specializing_iff_stable_under_specialization_image :
+  specializing f ↔ ∀ s, stable_under_specialization s → stable_under_specialization (f '' s) :=
+relation.fibration_iff_is_lower_set_image
+
+lemma specializing_iff_closure_singleton (hf : continuous f) :
+  specializing f ↔ ∀ x, f '' closure {x} = closure {f x} :=
+by simpa only [closure_singleton_eq_Iic] using
+  relation.fibration_iff_image_Iic hf.specialization_monotone
+
+lemma specializing_iff_is_closed_image_closure_singleton (hf : continuous f) :
+  specializing f ↔ ∀ x, is_closed (f '' closure {x}) :=
+begin
+  refine ⟨λ h x, _, λ h, specializing_iff_image_singleton_stable_under_specialization.mpr
+    (λ x, (h x).stable_under_specialization)⟩,
+  rw (specializing_iff_closure_singleton hf).mp h x,
+  exact is_closed_closure
+end
+
+lemma is_closed_map.specializing (hf : is_closed_map f) : specializing f :=
+specializing_iff_image_singleton_stable_under_specialization.mpr $
+  λ x, (hf _ is_closed_closure).stable_under_specialization
+
+lemma inducing.specializing (hf : inducing f) (h : stable_under_specialization (range f)) :
+  specializing f :=
+by { intros x y e, obtain ⟨y, rfl⟩ := h e ⟨x, rfl⟩, exact ⟨_, hf.specializes_iff.mp e, rfl⟩ }
+
+lemma inducing.generalizing (hf : inducing f) (h : stable_under_generalization (range f)) :
+  «generalizing» f :=
```

Comment:
Didn't realize "generalizing" is a keyword ... looks like it's declared [here](https://github.com/leanprover-community/lean/blob/db98bf983ba3da2cf0756a695e50242a27ddb5f0/library/init/meta/interactive.lean#L548)?